Bluetooth: Mesh: Add extra flag to control seg msg in frnd queue

PavelVPV opened the pull request:
This PR fixes issue introduced in
https://github.com/zephyrproject-rtos/zephyr/commit/5d059117fd279ee0eec4fb1c4d0244a37937013c.

Use extra flag stored in user data of net_buf to control segmented
messages in Friend Queue. The initial idea with using fragments didn't
work.

This fixes the following PTS tests:
- MESH/NODE/FRND/FN/BV-08-C
- MESH/NODE/FRND/FN/BV-19-C
- MESH/NODE/CFG/HBS/BV-05-C

PavelVPV committed:
Bluetooth: Mesh: Add extra flag to control seg msg in frnd queue

This commit fixes issue introduced in
5d059117fd279ee0eec4fb1c4d0244a37937013c.

Use extra flag stored in user data of net_buf to control segmented
messages in Friend Queue. The initial idea with using fragments didn't
work.

This fixes the following PTS tests:
- MESH/NODE/FRND/FN/BV-08-C
- MESH/NODE/FRND/FN/BV-19-C
- MESH/NODE/CFG/HBS/BV-05-C

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

PavelVPV committed:
tests: bluetooth: mesh: extend friend queue overflow test

Extend Friend Queue overflow test by overflow with segmented messages:
- when a segmented message is preempted by an unsegmented message;
- when a segmented message is preempted by another segmented message;

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

PavelVPV committed:
Revert "bluetooth: mesh: Remove illegal use of NET_BUF_FRAG in friend.c"

This reverts commit 5d059117fd279ee0eec4fb1c4d0244a37937013c.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

PavelVPV commented:
@carlescufi is it fine to have revert commit that won't compile on it's own (needs the next commit) or should I squash it with the next commit?

carlescufi commented:
> @carlescufi is it fine to have revert commit that won't compile on it's own (needs the next commit) or should I squash it with the next commit?

Please keep the revert commit even if it doesn't compile @PavelVPV.

alxelax commented:
@PavelVPV, could you add a bit more details why initial idea doesn't work?

PavelVPV commented:
> @PavelVPV, could you add a bit more details why initial idea doesn't work? 

My misunderstanding of the fragments feature. Is this enough?

alxelax commented on tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_friendship.c:
```diff
@@ -616,6 +662,65 @@ static void test_lpn_overflow(void)
+	ASSERT_EQUAL(exp_seq, msg.seq);
+
+	ASSERT_OK(bt_mesh_test_recv_msg(&msg, K_SECONDS(2)),
+		  "Receive the second unseg msg failed");
```
FYI This macro doesn't print text message. This is bug of ASSERT_OK.

PavelVPV replied:
I added analogue of `ASSERT_OK` macro that will print message: https://github.com/zephyrproject-rtos/zephyr/pull/53143/commits/49fee4f6db15373e70ba76bc59eebfbdb08ebc7f

PavelVPV committed:
tests: bluetooth: mesh: Add separate assert macro with message support

ASSERT_OK macro is confusing as it accepts additional arguments, but
don't use them. This commit removes the confusion by making ASSERT_OK
macro accept only one argument. An additional macro ASSERT_OK_MSG is
added that mandates a format string that will be printed and optionally
arguments for the format string.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

Andrewpini commented on tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_friendship.c:
```diff
@@ -303,8 +305,49 @@ static void test_friend_overflow(void)
+
+	LOG_INF("Testing overflow with unsegmented message preempting segmented one...");
+
+	/* Give the space in Friend Queue for one unsegmented message. */
```
```suggestion
	/* Make room in the Friend Queue for only one unsegmented message. */
```
Does this have the right meaning. The original comment is a little ambigious to me

Andrewpini commented on tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_friendship.c:
```diff
@@ -303,8 +305,49 @@ static void test_friend_overflow(void)
+	ASSERT_OK_MSG(evt_wait(FRIEND_POLLED, K_SECONDS(35)),
+		      "Friend never polled");
+
+	/* LPN verifies that no more messages are in Friend Queue. */
```
```suggestion
	/* LPN verifies that there are no more messages in the Friend Queue. */
```

Andrewpini commented on tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_friendship.c:
```diff
@@ -303,8 +305,49 @@ static void test_friend_overflow(void)
+
+	LOG_INF("Testing overflow with segmented message preempting another segmented...");
+
+	/* Give the space in Friend Queue for 2 unsegmented messages so the next unsgemented
```
```suggestion
	/* Make space in the Friend Queue for only 2 unsegmented messages so the next unsgemented
```

Andrewpini commented on tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_friendship.c:
```diff
@@ -616,6 +656,65 @@ static void test_lpn_overflow(void)
+	exp_seq += 2; /* Skipping the first message in Friend Queue. */
+	ASSERT_OK_MSG(bt_mesh_test_recv_msg(&msg, K_SECONDS(2)),
+		      "Receive first unseg msg failed");
+	ASSERT_EQUAL(5, msg.len);
+	ASSERT_EQUAL(cfg->addr, msg.ctx.recv_dst);
+	ASSERT_EQUAL(exp_seq, msg.seq);
```
Nitpick: This check is repeated 4-5 times, so it could be converted to a helper function. Up to you

Andrewpini reviewed (changes requested):
As far as I can see this looks good. Got a couple of comments regarding the added test coverage.
